src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,20 +126,54 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
     }
 
     qint64 offset = 0;
+    QByteArray protrudingData;
 
     QObject::connect(&socket, &QLocalSocket::readyRead, &loop, [&] {
-        auto data = socket.readAll();
-        if (data.isEmpty()) {
+        const auto dataReceived = socket.readAll();
+        if (dataReceived.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        // CFAPI expects sent blocks to be of a multiple of a block size.
+        // Only the last sent block is allowed to be of a different size than
+        // a multiple of a block size
+        constexpr auto cfapiBlockSize = 4096;
+
+        const QByteArray data = protrudingData + dataReceived;
+        protrudingData.clear();
+
+        QByteArray dataToSend;
+        // Extract blocks from data if possible
+        const auto blocksToSendCount = (data.size() - 1) / cfapiBlockSize;
+        const auto bytesToSendSize = cfapiBlockSize * blocksToSendCount;
+        if (bytesToSendSize != 0) {
+            dataToSend.append(data.left(bytesToSendSize));
+        }
+
+        // Do we still have any protruding data left? If yes, save this data for the next pass and send it then
+        if (bytesToSendSize < data.size()) {
+            protrudingData = data.right(data.size() - bytesToSendSize);
+            qDebug(lcCfApiWrapper) << "Save protruding data for next run. Size:" << protrudingData.size();
+        }
+
+        // Did we got all the data for the file? If yes, also append the protruding data
+        // that we saved for the next pass because this function will not be called again
+        if (!protrudingData.isEmpty() && callbackInfo->FileSize.QuadPart == offset + dataToSend.size() + protrudingData.size()) {
```

Comment:
@FlexW 
I am seeing a problem here. You are appending the leftover that did not fit the multiples of chunks to the chunked data. So, let's say your `dataToSend.size()` is **16384**. And, your `protrudingData.size()` is **512**. Appending the `protrudingData` to a `dataToSend` would then give you `dataToSend.size()` of **16384** + **512** which is **16896** which is not a multiple of **4096**. This is what will get pumped then into `cfExecute` - and, will probably fail? Only the very last part of a file is allowed to be not a multiple of **4096** right?
In this case, if, what I say above is right, you need to move this condition after

`// Finally send the data
        qDebug(lcCfApiWrapper) << "Send data block. Size:" << dataToSend.size();
        sendTransferInfo(dataToSend, offset);
        offset += dataToSend.size();`
        
        and then only check for `if (!protrudingData.isEmpty() && callbackInfo->FileSize.QuadPart == offset + protrudingData.size())` - only the `protrudingData` is missing, then - you can assume it's the last call to `readyRead` and you will now need to call `sendTransferInfo(protrudingData, offset);` or `dataToSend = protrudingData; sendTransferInfo(dataToSend, offset);`

Sorry, if I have missed something, but, that's what I understand from reading the code.

Reply:
Sorry, but I still think this code is too complicates for what it does and @allexzander comments show it. If we start from line 145 here, which currently is: `const QByteArray data = protrudingData + dataReceived;`

we can change it for:
```cpp
QByteArray data = protrudingData + dataReceived;
if (data.size() < cfapiBlockSize) {
    protudingData = data;
    return;
}
const auto protudingCount = data % cfapiBlockSize;
protudingData = data.right(protudingCount);
data.chop(protudingCount);
```

And now you know that `protudingData` got what didn't fit in n * 4096 and that `data` got exactly n * 4096 bytes and can be passed to `sendTransferInfo` like before.

That's what I tried to explain in one of my earlier comments but I thought I'd type it out this time. :-)

Reply:
@allexzander I think the logic is correct: The if statement only evaluates to true if this is the last pass. If it is the last pass, we append the protruding data (that we saved for the next pass) to dataToSend. Now, dataToSend might not be a multiple of 4096 but that's okay because it is the last pass. 
I'm sorry if I'm incorrect and miss something, but I now looked for a long time on the code and still can not see a mistake. I also tested the code a lot and didn't receive any errors.

@er-vin Thanks for writing it out:) It's sometimes tricky to grasp it just from words. With what you suggest, I see one minor problem: Let's say we are in the last run and `data.size()` is smaller than the block size. Then this protruding data would never be sent because of the return statement. We could work around this by adding the conditional that made @allexzander problems while reading (and which at the moment I think is correct). The only other escape from this situation would be to flush the data in the finished signal, but as I said before, we moved this flush logic out of the finish signal handler because sometimes the finish signal handler gets processed before the socket has read all data. The `readyRead()` function might sometimes be later processed as the finish signal because we call `loop.quit()` to late, but as I said, I'm not sure about it.
I was unsure if calls with `0` to `left()` and `right()` are defined, but if it's the case, we could compress my logic to the following. Then we wouldn't need the modulo operator (which always find somewhat confusing) and also would not misuse the `protrudingData` variable for a short period of time (that's maybe debatable if this is a misuse;)).

```cpp
const auto blocksToSendCount = data.size() / cfapiBlockSize;
const auto bytesToSendSize = cfapiBlockSize * blocksToSendCount;
dataToSend.append(data.left(bytesToSendSize));
protrudingData = data.right(std::min(0, data.size() - bytesToSendSize));

if (callbackInfo->FileSize.QuadPart == offset + dataToSend.size() + protrudingData.size()) {
    dataToSend.append(protrudingData);
}

sendTransferInfo(dataToSend, offset);
offset += dataToSend.size();
```
But of course I will happily go for the modulo version if you both like it more:)

Reply:
> @er-vin Thanks for writing it out:) It's sometimes tricky to grasp it just from words. With what you suggest, I see one minor problem: Let's say we are in the last run and `data.size()` is smaller than the block size. Then this protruding data would never be sent because of the return statement. We could work around this by adding the conditional that made @allexzander problems while reading (and which at the moment I think is correct). The only other escape from this situation would be to flush the data in the finished signal, but as I said before, we moved this flush logic out of the finish signal handler because sometimes the finish signal handler gets processed before the socket has read all data. The `readyRead()` function might sometimes be later processed as the finish signal because we call `loop.quit()` to late, but as I said, I'm not sure about it.

This is why I also said earlier this flushing should be done *after* the `loop.exec()` (there you know that no further signal can arrive). Might requires just some wrangling around the error handling case but overall this should make all of this way more readable with a more linear execution.

> Then we wouldn't need the modulo operator (which always find somewhat confusing)

This is unfortunate because that really makes it all way more readable and less error prone. Definition of modulo is *really* give me what's left after `n * blockSize`. It matches perfectly well here, otherwise you're bound to reinvent it by hand which makes for gratuitously more complicated code.

Reply:
@FlexW I actually never thought about modulo operator in the current context. Shame on me. Now, I see the code as more readable when using this modulo. Actually, I've been having some feeling that the calculation of this multiple should somehow be simplified, and, I totally forgot about the wonderful modulo operator :)

Reply:
@FlexW The idea of sending the very last part of a file after the `loop.exec();` also sounds like something we could try. Of course, I like the idea of having it all in the `readyRead()` lambda, but, if we do the changes suggested by @er-vin with that early-out, we may be forced to add even more complexity to `readyRead()` lambda to process the case of last data chunk arrival. I don't like this. So, moving the finalization code that sends the **protrudingData** after the last data has arrived, after the `loop.exe();` could really be our only option to keep it simple.

Reply:
> @allexzander I think the logic is correct: The if statement only evaluates to true if this is the last pass. If it is the last pass, we append the protruding data (that we saved for the next pass) to dataToSend. Now, dataToSend might not be a multiple of 4096 but that's okay because it is the last pass.

@FlexW This may indeed be the case. Just, please, test it while having the debugger attached and make sure the last block succeeds with `cfExecute()` even though it's not a multiple of 4096.

Reply:
@er-vin @allexzander I updated the code. I hope it's now better. Thank you for helping me and make the code more readable:)